internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,153 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/xmidt-org/wrp-go/v3"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/wrpkit"
+)
+
+var (
+	ErrInvalidInput     = errors.New("invalid input")
+	ErrMisconfiguredQOS = errors.New("misconfigured QOS")
+)
+
+// Option is a functional option type for QOS.
+type Option interface {
+	apply(*Handler) error
+}
+
+type optionFunc func(*Handler) error
+
+func (f optionFunc) apply(c *Handler) error {
+	return f(c)
+}
+
+// Handler queues incoming messages and sends them to the next wrphandler
+type Handler struct {
+	next wrpkit.Handler
+	// queue for wrp messages
+	queue PriorityQueue
+
+	m sync.Mutex
+	// ingestQueue blocks if the queue ingestion is already in progress, i.e.: send
+	// as many queued messages as possible
+	ingestQueue chan bool
+	// shutdown shuts down the queue ingestion
+	shutdown context.CancelFunc
+}
+
+// New creates a new instance of the Handler struct.  The parameter next is the
+// handler that will be called and monitored for errors.
+func New(next websocket.Egress, opts ...Option) (*Handler, error) {
+	if next == nil {
+		return nil, ErrInvalidInput
+	}
+
+	h := &Handler{
+		next:        next,
+		ingestQueue: make(chan bool, 1),
+	}
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt.apply(h); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return h, nil
+}
+
+// Start starts the qos queue ingestion.
+func (h *Handler) Start() {
+	h.m.Lock()
+	if h.shutdown != nil {
+		return
```

Comment:
This will deadlock the code because you exited the code block without releasing the mutex.